### ISSUE TITLE
ElasticSearch Tracing

### ISSUE THREAD

**Issue by wglanzer** [quarkusio/quarkus#3298]
**Description**
https://www.elastic.co/de/blog/distributed-tracing-opentracing-and-elastic-apm

Maybe it is possible to implement tracing with elasticsearch, like the jaeger-module does currently?

**Comment by gsmet** [comment]
@objectiser any thought on that? I'm not sure we want to keep it.

**Comment by objectiser** [comment]
The application can currently trace Elasticsearch explicitly, using https://github.com/opentracing-contrib/java-elasticsearch-client.

I guess this issue is more about implicitly being able to instrument the Elasticsearch client. Not sure this is currently possible.

**Comment by loicmathieu** [comment]
@objectiser if the issue is about using https://github.com/opentracing-contrib/java-elasticsearch-client then it's just about testing it and if it works adding something in the guide :)

That's what I did for JDBC tracing, you can see the section of the documentation here: https://quarkus.io/guides/opentracing#additional-instrumentation

**Comment by machi1990** [comment]
> @objectiser if the issue is about using https://github.com/opentracing-contrib/java-elasticsearch-client then it's just about testing it and if it works adding something in the guide :)
> 
> That's what I did for JDBC tracing, you can see the section of the documentation here: https://quarkus.io/guides/opentracing#additional-instrumentation

While at it, it will be awesome to  check that it works in native mode.

**Comment by loicmathieu** [comment]
@machi1990 yes, I was a little less explicit of the works that needs to be done:

- Test the dependency in both native and JVM mode
- Include it inside the bom, maybe you will need to fix some dependency issue as we use an old opentracing-api version ...
- Document it

**Comment by loicmathieu** [comment]
But in fact, if it's what @objectiser wants it's a different issue as this one that is related to Elasticsearch APM instead of Jaeger, so maybe the best is to open a different issue. 
So @objectiser if it's what you want, please open a different issue for this with the steps to follow to fix it (see my previous comment).
If you want to take care of it, please proceed with it, otherwise I can take care of it easily :)

**Comment by objectiser** [comment]
@loicmathieu I am unclear what the issue author was asking - but if they want implicit tracing then I don't think that is possible currently - so the ticket should be closed.

Regarding your suggestion about adding to docs - we need to be careful with this, as there are a wide range of instrumentations available in opentracing-contrib. Are we intending to add them all here? How will they be tested, as I don't think we want to be manually testing them with each release?

**Comment by loicmathieu** [comment]
@objectiser what we do to JDBC tracing is just manual testing in both JVM and native, documentation and dependency management. We didn't add any automated testing, nor did we add them automatically as dependencies.

And, as we didn't add any automatic test, nor do any manual testing, there is some (very low) risk that this didn't works in some point after some changes done on Quarkus/OpenTracing/... so maybe we should add a note about this in the documentation.

**Comment by objectiser** [comment]
> so maybe we should add a note about this in the documentation.

@loicmathieu Yes - and it also might be worth defining the version of Quarkus it was last tested with?

**Comment by loicmathieu** [comment]
> @loicmathieu Yes - and it also might be worth defining the version of Quarkus it was last tested with?

I don't think the version adds any information as it's support will depends on the changes on native/opentracing/quarkus ... and some user uses it on production and make upgrade (like me, I use it without any issue with Quarkus 0.28.1).

**Comment by wglanzer** [comment]
The idea behind this ticket was that I wanted to use the Elastic-Stack as a tracing server / performance data receiver.
They offer a product called APM ( https://www.elastic.co/products/apm ) and if you install an elasticsearch server, you get APM as well.

**Comment by objectiser** [comment]
@wglanzer Ah ok - sorry totally misunderstood. I guess a question for the maintainers, but if you are interested in submitting a PR I think it would be a good addition.

**Comment by IRQs** [comment]
The integration of Quarkus OpenTracing with Elastic APM does not seem to be straightforward due to two reasons:
1. The (Elastic) APM Agent focuses primarily on [Servlet-based APIs][1] and does not support Vert.x. I have not tested Quarkus' support for Servlets via Undertow though. 
2. Ideally, one could use the [Elastic APM OpenTracing Bridge][2] by switching the tracer implementation. Unfortunately, switching tracers is not supported at the moment. In addition, the upstream project Smallrye OpenTracing is [stuck on OpenTracing API 0.31][3] whereas the Elastic OpenTracing Bridge is built against 0.33.

Having said this, I was able to use the new [Jaeger integration available with Elastic APM 7.6][4], since Quarkus OpenTracing is built with Jaeger internally. Beware though, that this is an experimental feature.

[1]: https://www.elastic.co/guide/en/apm/agent/java/1.x/supported-technologies-details.html#supported-web-frameworks
[2]: https://www.elastic.co/guide/en/apm/agent/java/1.x/opentracing-bridge.html
[3]: https://github.com/smallrye/smallrye-opentracing/issues/85
[4]: https://www.elastic.co/guide/en/apm/server/7.6/jaeger.html

**Comment by nadworny** [comment]
@IRQs could you please provide some details on how did you implement it using Jaeger integration?
